Material inspection

PepperCode1 opened the pull request:
- Add `MaterialView` to retrieve material properties from `RenderMaterial` and `MaterialFinder`
- Add `MaterialFinder#copyFrom(MaterialView)` to copy properties from another `MaterialView`
- Finish todo in `QuadView#toBakedQuad`
- Move material impl classes to `material` package
- Improve material properties' javadoc by explaining which contexts support certain properties
- Add `glint` render material property
- Add `MutableQuadView#copyFrom` and deprecate `QuadView#copyTo`

PepperCode1 committed:
Material inspection

- Add MaterialView to get material properties from RenderMaterial and MaterialFinder
- Add MaterialFinder#copyFrom to copy properties from another MaterialView
- Finish todo in QuadView#toBakedQuad
- Move material impl classes to material package

PepperCode1 committed:
Add glint material property

- Allow force enabling or force disabling glint on items
- Force enable glint on horizontal sides of pillar item model in test mod

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialFinder.java:
```diff
@@ -92,6 +92,26 @@ default MaterialFinder ambientOcclusion(TriState mode) {
 	}
 
+	/**
+	 * @apiNote The default implementation will be removed in the next breaking release.
```
Needs documentation explaining what this does. (Similar to `ambientOcclusion` which also uses a `TriState`).

Technici4n replied:
It's quite important to mention that this property only works in item contexts, and will be entirely ignored in block contexts.

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialFinder.java:
```diff
@@ -92,6 +92,26 @@ default MaterialFinder ambientOcclusion(TriState mode) {
+	}
+
+	/**
+	 * @apiNote The default implementation will be removed in the next breaking release.
```
Needs documentation too.

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialView.java:
```diff
@@ -0,0 +1,66 @@
+import net.fabricmc.fabric.api.util.TriState;
+
+/**
+ * Getter methods for {@link RenderMaterial} (immutable) and {@link MaterialFinder} (mutable).
```
This must mention that the properties may not match exactly what was requested depending on the renderer implementation. Something like https://discord.com/channels/507304429255393322/566276937035546624/951272466116915290.

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialView.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Getter methods for {@link RenderMaterial} (immutable) and {@link MaterialFinder} (mutable).
+ */
+public interface MaterialView {
```
We can link the `MaterialFinder` functions as javadoc to explain what the methods do. I think a simple `@see MaterialFinder#relevantMethod` would do.

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/model/FabricBakedModel.java:
```diff
@@ -25,8 +25,8 @@
 import net.minecraft.client.util.math.MatrixStack;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.math.BlockPos;
-import net.minecraft.world.BlockRenderView;
```
Should not reformat imports in unrelated files.

Technici4n commented on fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/material/RenderMaterialImpl.java:
```diff
@@ -0,0 +1,96 @@
+		}
+	}
+
+	private final BlendMode blendMode;
```
This increases the size of materials quite a bit. Is it really needed? It wasn't before...

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialFinder.java:
```diff
@@ -92,6 +92,26 @@ default MaterialFinder ambientOcclusion(TriState mode) {
+	/**
+	 * @apiNote The default implementation will be removed in the next breaking release.
+	 */
+	default MaterialFinder copyFrom(MaterialView material) {
```
Remember to also add `copyFrom` in `MutableQuadView` as discussed.

Technici4n reviewed (changes requested):
First batch of comments.

PepperCode1 committed:
Fix imports

PepperCode1 committed:
Add documentation

PepperCode1 committed:
Do not create invalid materials

- Ensure material bits are valid before creating material
- Assert that default MaterialFinder bits are valid
- Remove ordinal checks in MaterialViewImpl since bits are now guaranteed to be valid
- Set default glint mode to default instead of false
- Remove getter overrides in RenderMaterialImpl

PepperCode1 committed:
Add MutableQuadView#copyFrom

- Deprecate QuadView#copyTo

PepperCode1 committed:
Add missing nullability annotations

Technici4n reviewed (approved):
Perfect, thank you!

The following uses of `RenderMaterialImpl` can be replaced by `RenderMaterial` if you would like to, I leave that up to you:
- return value of `EncodingFormat#material(int)`
- return value of `QuadViewImpl#material()`
- type of `RenderMaterialImpl#BY_INDEX`
- return value of `RenderMaterial#byIndex`
- return value of `RenderMaterial#setDisableDiffuse`

EDIT: As discussed on Discord, these changes will not be applied.
